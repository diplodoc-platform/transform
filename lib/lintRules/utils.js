const {LOG_LEVELS, LOG_LEVELS_PRIORITIES} = require('../constants');

function isSupportedLevel(level) {
    return Object.values(LOG_LEVELS).includes(level);
}

function compareLogLevels(levelA, levelB) {
    return LOG_LEVELS_PRIORITIES[levelA] - LOG_LEVELS_PRIORITIES[levelB];
}

function withLintRuleProperties(wrapper, lintRule) {
    wrapper.ruleName = lintRule.ruleName;
    wrapper.ruleOptions = lintRule.ruleOptions;

    return wrapper;
}

function withCheckDisabledRule(lintRule) {
    function wrapper(params) {
        const {ruleOptions} = params;

        if (ruleOptions.disabled) {
            return;
        }

        lintRule(params);
    }

    return withLintRuleProperties(wrapper, lintRule);
}

function getCountLogMessages(log, level) {
    return log.get()[level].length;
}

function withValidationLevelOptions(lintRule) {
    function wrapper(params) {
        const {ruleOptions, commonOptions: {log, path}} = params;
        const {level} = ruleOptions;

        if (isSupportedLevel(level)) {
            const messageCount = getCountLogMessages(log, level);

            lintRule(params);

            if (getCountLogMessages(log, level) > messageCount &&
                compareLogLevels(lintRule.ruleOptions.level, level) > 0
            ) {
                log.warn(`The log level of lint rule "${lintRule.ruleName}" was decreased. ` +
                    `This can lead to problems with content in ${path}`);
            }
        } else {
            lintRule({...params, ruleOptions: {
                ...ruleOptions,
                level: lintRule.ruleOptions.level,
            }});
        }
    }

    return withLintRuleProperties(wrapper, lintRule);
}

function withDefaultOptions(lintRule) {
    function wrapper(params) {
        const {ruleOptions: customRuleOptions} = params;
        const defaultRuleOptions = lintRule.ruleOptions || {};
        const ruleOptions = {...defaultRuleOptions, ...customRuleOptions};

        lintRule({...params, ruleOptions});
    }

    return withLintRuleProperties(wrapper, lintRule);
}

function withOwnOptions(lintRule) {
    function wrapper(params) {
        const {lintOptions = {}, commonOptions = {}} = params;

        if (commonOptions.disableLint) {
            return;
        }

        const ruleOptions = lintOptions[lintRule.ruleName] || {};

        lintRule({...params, ruleOptions});
    }

    return withLintRuleProperties(wrapper, lintRule);
}

function withLintRuleWrapper(lintRule) {
    return withOwnOptions(
        withDefaultOptions(
            withCheckDisabledRule(
                withValidationLevelOptions(
                    lintRule,
                ))));
}

module.exports = {
    withLintRuleWrapper,
    isSupportedLevel,
    compareLogLevels,
};


