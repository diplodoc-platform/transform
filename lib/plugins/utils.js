const {bold} = require('chalk');
const {platform} = require('process');

const nestedCloseTokenIdxFactory = (tokenName, matchOpenToken, matchCloseToken) => (tokens, idx, path, log) => {
    let level = 0;
    let i = idx;
    while (i < tokens.length) {
        if (matchOpenToken(tokens, i)) {
            level++;
        } else if (matchCloseToken(tokens, i)) {
            if (level === 0) {
                return i;
            }
            level--;
        }

        i++;
    }

    log.error(`${tokenName} must be closed${path ? ` in ${bold(path)}` : ''}`);

    return null;
};

const сarriage = platform === 'win32' ? '\r\n' : '\n';

module.exports = {nestedCloseTokenIdxFactory, сarriage};
