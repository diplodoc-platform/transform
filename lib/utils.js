function isLocalUrl(url) {
    return !/^(?:[a-z]+:)?\/\//i.test(url);
}

function findBlockTokens(tokens, id) {
    let blockTokens = [];
    let i = 0, startToken, start, end;
    while (i < tokens.length) {
        const token = tokens[i];

        if (typeof start === 'number') {
            if (startToken.type === 'paragraph_open' && token.type === 'paragraph_close') {
                end = i + 1;
                break;
            } else if (startToken.type === 'heading_open') {
                if (token.type === 'heading_open' && token.tag === startToken.tag) {
                    end = i;
                    break;
                } else if (i === tokens.length - 1) {
                    end = tokens.length;
                }
            }
        }

        if (
            (token.type === 'paragraph_open' || token.type === 'heading_open') &&
            token.attrGet('id') === id &&
            typeof start === 'undefined'
        ) {
            startToken = token;
            start = i;
        }

        i++;
    }

    if (typeof start === 'number' && typeof end === 'number') {
        blockTokens = tokens.slice(start, end);
    }

    return blockTokens;
}

function headingInfo(tokens, idx) {
    const openToken = tokens[idx];
    const inlineToken = tokens[idx + 1];

    let lastTextToken, i = 0;
    while (i < inlineToken.children.length) {
        const token = inlineToken.children[i];

        if (token.type === 'text') {
            lastTextToken = token;
        }

        i++;
    }

    const level = Number.parseInt(openToken.tag.slice(1), 10);
    const title = lastTextToken && lastTextToken.content || inlineToken.content;

    return {
        level,
        title,
    };
}

function isExternalHref(href) {
    return href.startsWith('http') || href.startsWith('//');
}

function transformLinkToOriginalArticle({root, currentPath}) {
    return currentPath
        .replace(root, '')
        .replace(/\.(md|ya?ml|html)$/i, '');
}

module.exports = {
    isLocalUrl,
    findBlockTokens,
    headingInfo,
    isExternalHref,
    transformLinkToOriginalArticle,
};
