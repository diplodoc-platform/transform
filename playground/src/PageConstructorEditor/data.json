{
  "blocks": [
    {
      "type": "header-block",
      "width": "xs",
      "offset": "default",
      "title": "Как всё устроено?",
      "resetPaddings": true,
      "verticalOffset": "l",
      "description": "Diplodoc — это платформа документации со следущими компонентами:\n* клиентская часть отвечает за отображение документов в браузере на стороне пользователя;\n* серверная часть отвечает за хостинг и выдачу документов клиенту;\n* инструменты создания и преобразования документов из Markdown в статический сайт (Transform & Build).\n\nСистема выполняет основные операции на Node.js и React.",
      "background": {
        "image": {
          "src": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-cover.png",
          "disableCompress": true
        }
      }
    },
    {
      "type": "extended-features-block",
      "title": {
        "text": "Подход Docs as Сode"
      },
      "colSizes": {
        "all": 12,
        "xl": 3,
        "md": 4,
        "sm": 6
      },
      "items": [
        {
          "title": "Markdown",
          "icon": "https://storage.yandexcloud.net/cloud-www-assets/pages/how-to-diplodoc/how-to-diplodoc-01.svg",
          "text": "Документация пишется на языке разметки Markdown, а не в plain text или WYSIWYG-редакторе."
        },
        {
          "title": "Хранение в репозитории",
          "icon": "https://storage.yandexcloud.net/cloud-www-assets/pages/how-to-diplodoc/how-to-diplodoc-02.svg",
          "text": "Документация хранится в репозитории с остальным кодом. Это может быть GitHub или ваше собственное решение. "
        },
        {
          "title": "Налаженный процесс",
          "icon": "https://storage.yandexcloud.net/cloud-www-assets/pages/how-to-diplodoc/how-to-diplodoc-03.svg",
          "text": "Документация собирается в финальный проект при помощи генератора статических сайтов и публикуется автоматически."
        },
        {
          "title": "Синхронизация",
          "icon": "https://storage.yandexcloud.net/cloud-www-assets/pages/how-to-diplodoc/how-to-diplodoc-04.svg",
          "text": "Документация пишется и обновляется совместно — точно так же, как и остальной код."
        }
      ]
    },
    {
      "type": "card-layout-block",
      "title": "Хранение данных",
      "children": [
        {
          "type": "layout-item",
          "content": {
            "title": "Добавление",
            "text": "Исходный код документации хранится в репозитории клиента в формате Markdown."
          },
          "media": {
            "image": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-item-02-01.png"
          }
        },
        {
          "type": "layout-item",
          "content": {
            "title": "Преобразование",
            "text": "Преобразованный из Markdown-файлов документационный проект попадает в отдельную директорию S3-совместимого хранилища для каждого клиента."
          },
          "media": {
            "image": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-item-02-02.png"
          }
        },
        {
          "type": "layout-item",
          "content": {
            "title": "Доступ",
            "text": "Доступ к S3 регулируется с помощью уникальных ключей."
          },
          "media": {
            "image": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-item-02-03.png"
          }
        }
      ]
    },
    {
      "type": "card-layout-block",
      "title": "Развёртывание",
      "colSizes": {
        "all": 12,
        "xl": 3,
        "md": 4,
        "sm": 6
      },
      "children": [
        {
          "type": "layout-item",
          "content": {
            "title": "Хранение",
            "text": "Документационный проект из S3-совместимого хранилища выкладывается на Diplodoc.com."
          },
          "media": {
            "image": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-item-03-01.png"
          }
        },
        {
          "type": "layout-item",
          "content": {
            "title": "Проксирование",
            "text": "Возможно проксирование с сайта клиента на сайт diplodoc.com для бесшовной интеграции."
          },
          "media": {
            "image": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-item-03-02.png"
          }
        },
        {
          "type": "layout-item",
          "content": {
            "title": "Поиск",
            "text": "Поддержка поиска с помощью Elasticsearch.  "
          },
          "media": {
            "image": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-item-03-03.png"
          }
        },
        {
          "type": "layout-item",
          "content": {
            "title": "Локализация",
            "text": "Поддержка локализации и кастомизации проекта документации. "
          },
          "media": {
            "image": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-item-03-04.png"
          }
        }
      ]
    },
    {
      "type": "card-layout-block",
      "title": "Полная совместимость",
      "children": [
        {
          "type": "background-card",
          "title": "YFM",
          "text": "Диалект Markdown, соответствующий спецификации CommonMark, с поддержкой плагинов и дополнительными возможностями разметки.",
          "background": {
            "src": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-card-01.png",
            "disableCompress": true
          },
          "paddingBottom": "m",
          "border": "none"
        },
        {
          "type": "background-card",
          "title": "GitHub",
          "text": "Полная интеграция с системами контроля версий, в первую очередь, с GitHub — одним из основных инструментов разработки.",
          "background": {
            "src": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-card-02.png",
            "disableCompress": true
          },
          "paddingBottom": "m",
          "border": "none"
        },
        {
          "type": "background-card",
          "title": "Elasticsearch",
          "text": "Быстрая индексация и поиск на основе Elasticsearch для обеспечения релевантности и точности результатов запросов.",
          "background": {
            "src": "https://storage.yandexcloud.net/diplodoc-www-assets/pages/how-to-diplodoc/ddos-how-card-03.png",
            "disableCompress": true
          },
          "paddingBottom": "m",
          "border": "none"
        }
      ]
    },
    {
      "type": "icons-block",
      "size": "m",
      "title": "Присоединяйтесь к сообществу",
      "items": [
        {
          "src": "https://storage.yandexcloud.net/cloud-www-assets/pages/index-diplodoc/github.svg",
          "text": "GitHub",
          "url": "#"
        },
        {
          "src": "https://storage.yandexcloud.net/cloud-www-assets/pages/index-diplodoc/telegram.svg",
          "text": "Telegram",
          "url": "#"
        },
        {
          "src": "https://storage.yandexcloud.net/cloud-www-assets/pages/index-diplodoc/stackoverflow.svg",
          "text": "Stack Overflow",
          "url": "#"
        }
      ]
    }
  ]
}