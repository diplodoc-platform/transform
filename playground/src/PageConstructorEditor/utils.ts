import isEqual from 'lodash/isEqual';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const memoizeLast = (fn: (...args: any[]) => any) => {
  let cacheKey: Parameters<typeof fn>;
  let cacheResult: ReturnType<typeof fn>;

  return (...params: Parameters<typeof fn>) => {
    if (!isEqual(params, cacheKey)) {
      try {
        const result = fn(...params);

        cacheResult = result;
        cacheKey = params;
      } catch {
        return cacheResult;
      }
    }

    return cacheResult;
  };
};
