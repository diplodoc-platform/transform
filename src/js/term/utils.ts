export const Selector = {
    TITLE: '.yfm .yfm-term_title',
    CONTENT: '.yfm .yfm-term_dfn',
};
export const openClass = 'open';
export const openDefinitionClass = Selector.CONTENT.replace(/\./g, '') + ' ' + openClass;
let isListenerNeeded = true;

export function createDefinitionElement(termElement: HTMLElement) {
    const termKey = termElement.getAttribute('term-key');
    const definitionTemplate = document.getElementById(
        `${termKey}_template`,
    ) as HTMLTemplateElement;
    const definitionElement = definitionTemplate?.content.cloneNode(true).firstChild as HTMLElement;

    definitionTemplate?.parentElement?.appendChild(definitionElement);
    definitionTemplate.remove();

    return definitionElement;
}

export function setDefinitionId(definitionElement: HTMLElement, termElement: HTMLElement): void {
    const termId = termElement.getAttribute('id') || Math.random().toString(36).substr(2, 8);
    definitionElement?.setAttribute('term-id', termId);
}

export function setDefinitionPosition(
    definitionElement: HTMLElement,
    termElement: HTMLElement,
): void {
    const {
        x: termX,
        y: termY,
        right: termRight,
        left: termLeft,
        width: termWidth,
        height: termHeight,
    } = termElement.getBoundingClientRect();

    const termParent = termParentElement(termElement);

    if (!termParent) {
        return;
    }

    const {right: termParentRight, left: termParentLeft} = termParent.getBoundingClientRect();

    if ((termParentRight < termLeft || termParentLeft > termRight) && !isListenerNeeded) {
        closeDefinition(definitionElement);
        return;
    }

    if (isListenerNeeded && termParent) {
        termParent.addEventListener('scroll', termOnResize);
        isListenerNeeded = false;
    }

    const relativeX = Number(definitionElement.getAttribute('relativeX'));
    const relativeY = Number(definitionElement.getAttribute('relativeY'));

    if (relativeX === termX && relativeY === termY) {
        return;
    }

    definitionElement.setAttribute('relativeX', String(termX));
    definitionElement.setAttribute('relativeY', String(termY));

    const offsetTop = termHeight + 5;
    const definitionParent = definitionElement.parentElement;

    if (!definitionParent) {
        return;
    }

    const {width: definitionWidth} = definitionElement.getBoundingClientRect();
    const {left: definitionParentLeft} = definitionParent.getBoundingClientRect();

    // If definition not fit document change base alignment
    const definitionRightCoordinate = definitionWidth + Number(getCoords(termElement).left);
    const fitDefinitionDocument =
        document.body.clientWidth > definitionRightCoordinate ? 0 : definitionWidth - termWidth;

    const customHeaderTop = getCoords(definitionParent).top - definitionParent.offsetTop;

    definitionElement.style.top =
        Number(getCoords(termElement).top + offsetTop - customHeaderTop) + 'px';
    definitionElement.style.left =
        Number(
            getCoords(termElement).left -
                definitionParentLeft +
                definitionParent.offsetLeft -
                fitDefinitionDocument,
        ) + 'px';
}

function termOnResize() {
    const openDefinition = document.getElementsByClassName(openDefinitionClass)[0] as HTMLElement;

    if (!openDefinition) {
        return;
    }
    const termId = openDefinition.getAttribute('term-id') || '';
    const termElement = document.getElementById(termId);

    if (!termElement) {
        return;
    }

    setDefinitionPosition(openDefinition, termElement);
}

function termParentElement(term: HTMLElement | null) {
    if (!term) {
        return null;
    }

    const closestScrollableParent = term.closest('table') || term.closest('code');

    return closestScrollableParent || term.parentElement;
}

export function closeDefinition(definition: HTMLElement) {
    definition.classList.remove(openClass);
    const termId = definition.getAttribute('term-id') || '';
    const termParent = termParentElement(document.getElementById(termId));

    if (!termParent) {
        return;
    }

    termParent.removeEventListener('scroll', termOnResize);
    isListenerNeeded = true;
}

function getCoords(elem: HTMLElement) {
    const box = elem.getBoundingClientRect();

    const body = document.body;
    const docEl = document.documentElement;

    const scrollTop = window.pageYOffset || docEl.scrollTop || body.scrollTop;
    const scrollLeft = window.pageXOffset || docEl.scrollLeft || body.scrollLeft;

    const clientTop = docEl.clientTop || body.clientTop || 0;
    const clientLeft = docEl.clientLeft || body.clientLeft || 0;

    const top = box.top + scrollTop - clientTop;
    const left = box.left + scrollLeft - clientLeft;

    return {top: Math.round(top), left: Math.round(left)};
}
