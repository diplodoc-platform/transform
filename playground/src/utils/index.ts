import {EditorType} from "src/App";
import data from 'src/PageConstructorEditor/data.json';
import initialMarkup from 'src/SplitViewEditor/mdContent';

function persist(input: string, query = 'input') {
  try {
    const current = new URL(location.toString());
    current.searchParams.set(query, encodeURI(input) ?? '');

    history.pushState(null, null, current);
  } catch (err) {
    console.error(err);
  }
}

function deleteQuery(query) {
  try {
    const current = new URL(location.toString());
    current.searchParams.delete(query);

    history.pushState(null, null, current);
  } catch (err) {
    console.error(err);
  }
}

function restore(query = 'input') {
  try {
    const search = new URLSearchParams(window.location.search);
    let input = search.get(query);

    if (input?.length) {
      input = decodeURI(input);
    } else {
      return '';
    }

    return input;
  } catch (err) {
    return '';
  }
}

function prefill() {
  const search = new URLSearchParams(window.location.search);

  if(search.get('prefill') === 'true'){
    const mode = search.get('mode')
    return mode === EditorType.PC ? data : initialMarkup
  }

  return restore();
}

function deleteElementById(data, idToRemove) {
  return data.map(subArray => {
    return subArray.filter(element => element.id !== idToRemove);
  }).filter(subArray => subArray.length > 0);
}


export {persist, restore, prefill, deleteQuery, deleteElementById};
