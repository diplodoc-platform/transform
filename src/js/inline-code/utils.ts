import {getCoords} from '../term/utils';

import {OPEN_CLASS} from './constant';

export let timer: ReturnType<typeof setTimeout> | null = null;

let isListenerNeeded = true;

export function getTooltipElement(): HTMLElement | null {
    return document.querySelector('.yfm-inline-code-tooltip');
}

function setTooltipAriaAttributes(tooltipElement: HTMLElement, targetElement: HTMLElement): void {
    const ariaLive = targetElement.getAttribute('aria-live') || 'polite';
    tooltipElement?.setAttribute('aria-live', ariaLive);
    tooltipElement?.setAttribute('aria-modal', 'true');
}

function checkTimerAndClear() {
    if (timer) {
        clearTimeout(timer);
        timer = null;
    }
}

function tooltipParentElement(target: HTMLElement | null) {
    if (!target) {
        return null;
    }

    const closestScrollableParent = target.closest('table') || target.closest('code');

    return closestScrollableParent || target.parentElement;
}

function tooltipOnResize() {
    const openedDefinition = getTooltipElement();

    if (!openedDefinition) {
        return;
    }
    const inlineId = openedDefinition.getAttribute('inline-id') || '';
    const targetElement = document.getElementById(inlineId);

    if (!targetElement) {
        return;
    }

    setTooltipPosition(openedDefinition, targetElement);
}

export function setTooltipPosition(tooltipElement: HTMLElement, targetElement: HTMLElement): void {
    const {
        x: inlineX,
        y: inlineY,
        right: inlineRight,
        left: inlineLeft,
        width: inlineWidth,
        height: inlineHeight,
    } = targetElement.getBoundingClientRect();

    const tooltipParent = tooltipParentElement(targetElement);

    if (!tooltipParent) {
        return;
    }

    const {right: tooltipParentRight, left: tooltipParentLeft} =
        tooltipParent.getBoundingClientRect();

    if ((tooltipParentRight < inlineLeft || tooltipParentLeft > inlineRight) && !isListenerNeeded) {
        closeTooltip(tooltipElement);
        return;
    }

    if (isListenerNeeded && tooltipParent) {
        tooltipParent.addEventListener('scroll', tooltipOnResize);
        isListenerNeeded = false;
    }

    const relativeX = Number(tooltipElement.getAttribute('relativeX'));
    const relativeY = Number(tooltipElement.getAttribute('relativeY'));

    if (relativeX === inlineX && relativeY === inlineY) {
        return;
    }

    tooltipElement.setAttribute('relativeX', String(inlineX));
    tooltipElement.setAttribute('relativeY', String(inlineY));

    const offsetTop = inlineHeight + 5;
    const definitionParent = tooltipElement.parentElement;

    if (!definitionParent) {
        return;
    }

    const {width: definitionWidth} = tooltipElement.getBoundingClientRect();
    const {left: definitionParentLeft} = definitionParent.getBoundingClientRect();

    // If definition not fit document change base alignment
    const definitionLeftCoordinate = Number(getCoords(targetElement).left);
    const definitionRightCoordinate = definitionWidth + definitionLeftCoordinate;

    const definitionOutOfScreenOnLeft = definitionLeftCoordinate - definitionWidth < 0;
    const definitionOutOfScreenOnRight = definitionRightCoordinate > document.body.clientWidth;

    const isAlignSwapped = definitionOutOfScreenOnRight || document.dir === 'rtl';
    const fitDefinitionDocument =
        isAlignSwapped && !definitionOutOfScreenOnLeft ? definitionWidth - inlineWidth : 0;
    const customHeaderTop = getCoords(definitionParent).top - definitionParent.offsetTop;
    const offsetRight = 5;
    const shiftLeft = definitionOutOfScreenOnRight
        ? definitionRightCoordinate - document.body.clientWidth + offsetRight
        : 0;
    const offsetLeft =
        getCoords(targetElement).left -
        definitionParentLeft +
        definitionParent.offsetLeft -
        fitDefinitionDocument;

    const isShiftLeftNeeded = offsetLeft + definitionWidth >= document.body.clientWidth;

    tooltipElement.style.top =
        Number(getCoords(targetElement).top + offsetTop - customHeaderTop) + 'px';
    tooltipElement.style.left = Number(offsetLeft - (isShiftLeftNeeded ? shiftLeft : 0)) + 'px';
}

export function getInlineCodeByTooltip(definition: HTMLElement) {
    const inlineId = definition.getAttribute('inline-id');

    return inlineId ? document.getElementById(inlineId) : null;
}

function closeTooltipFn(definition: HTMLElement) {
    definition.classList.remove(OPEN_CLASS);
    const inline = getInlineCodeByTooltip(definition);
    const tooltipParent = tooltipParentElement(inline);

    definition.removeAttribute('inline-id');

    if (!tooltipParent) {
        return;
    }

    tooltipParent.removeEventListener('scroll', tooltipOnResize);
    isListenerNeeded = true;
}

export function openTooltip(target: HTMLElement) {
    const tooltipElement = document.getElementById('tooltip_inline_clipboard_dialog');

    if (!target.matches('.yfm-clipboard-inline-code') || !tooltipElement) {
        return;
    }

    tooltipElement.setAttribute('inline-id', target.getAttribute('id') || '');
    setTooltipAriaAttributes(tooltipElement, target);
    setTooltipPosition(tooltipElement, target);

    // In order not to get rid of the smooth appearance effect, I had to do this
    if (tooltipElement.classList.contains(OPEN_CLASS)) {
        tooltipElement.classList.remove(OPEN_CLASS);
        requestAnimationFrame(() => {
            tooltipElement.classList.add(OPEN_CLASS);
        });
    } else {
        tooltipElement.classList.add(OPEN_CLASS);
    }

    return tooltipElement;
}

export function closeTooltip(target: HTMLElement) {
    checkTimerAndClear();
    closeTooltipFn(target);
}

export function tooltipWorker(target: HTMLElement) {
    const definition = openTooltip(target);

    if (!definition) {
        return;
    }
    checkTimerAndClear();
    timer = setTimeout(() => {
        closeTooltip(definition);
        timer = null;
    }, 1000);
}
