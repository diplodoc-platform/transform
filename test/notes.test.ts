import dedent from 'ts-dedent';

import alerts from '../src/transform/plugins/notes';
import transform from '../src/transform';

import {callPlugin, tokenize} from './utils';
import {base, customTitle, emptyTitle} from './data/alerts';

const html = (text: string, notesAutotitle = true) => {
    const {
        result: {html},
    } = transform(text, {
        notesAutotitle,
        plugins: [alerts],
    });
    return html;
};

describe('Alerts', () => {
    test('Should transform to new tokens', () => {
        const result = callPlugin(
            alerts,
            tokenize([
                'Text before',
                '',
                '{% note info %}',
                '',
                'Текст примечания.',
                '',
                '{% endnote %}',
                '',
                'Text after',
            ]),
        );

        expect(result).toEqual(base);
    });

    describe('Should support all types', () => {
        [
            {type: 'info', title: 'Примечание'},
            {type: 'tip', title: 'Совет'},
            {type: 'alert', title: 'Внимание'},
            {type: 'warning', title: 'Важно'},
        ].forEach(({type, title}) => {
            test(`should support type: ${type}`, () => {
                const result = callPlugin(
                    alerts,
                    tokenize([`{% note ${type} %}`, '', 'Текст примечания.', '', '{% endnote %}']),
                );

                expect(result[0].attrs?.[0][1]).toEqual(`yfm-note yfm-accent-${type}`);
                expect(result[2].children?.[0].content).toEqual(title);
            });
        });
    });

    test('Should support custom title', () => {
        const result = callPlugin(
            alerts,
            tokenize([
                'Text before',
                '',
                '{% note info "Custom title" %}',
                '',
                'Текст примечания.',
                '',
                '{% endnote %}',
                '',
                'Text after',
            ]),
        );
        expect(result).toEqual(customTitle);
    });

    test('Should support empty title', () => {
        const result = callPlugin(
            alerts,
            tokenize([
                'Text before',
                '',
                '{% note info "" %}',
                '',
                'Текст примечания.',
                '',
                '{% endnote %}',
                '',
                'Text after',
            ]),
        );

        expect(result).toEqual(emptyTitle);
    });

    test('should render simple note', () => {
        expect(
            html(dedent`
            {% note info "Note title" %}

            Note content

            {% endnote %}
        `),
        ).toMatchSnapshot();
    });

    test('should render siblings notes', () => {
        expect(
            html(dedent`
            {% note info "Note title 1" %}

            Note content 1

            {% endnote %}

            {% note info "Note title 2" %}

            Note content 2

            {% endnote %}
        `),
        ).toMatchSnapshot();
    });

    test('should render nested notes', () => {
        expect(
            html(dedent`
            {% note info "Outer title" %}

            Outer content

            {% note info "Inner title" %}

            Inner content

            {% endnote %}

            {% endnote %}
        `),
        ).toMatchSnapshot();
    });

    test('should render title with format', () => {
        expect(
            html(dedent`
            {% note info "_Italic note title_" %}

            Content

            {% endnote %}
        `),
        ).toMatchSnapshot();
    });

    test('should render note without title (notesAutotitle: true)', () => {
        expect(
            html(dedent`
                {% note info %}

                Note content without title

                {% endnote %}
            `),
        ).toMatchSnapshot();
    });

    test('should render note without title (notesAutotitle: false)', () => {
        expect(
            html(
                dedent`
                {% note info %}

                Note content without title

                {% endnote %}
            `,
                false,
            ),
        ).toMatchSnapshot();
    });

    test('should render note with empty string title', () => {
        expect(
            html(dedent`
                {% note info "" %}

                Note content with empty string title

                {% endnote %}
            `),
        ).toMatchSnapshot();
    });
});
