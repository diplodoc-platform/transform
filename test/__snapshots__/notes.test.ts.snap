// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alerts should render nested notes 1`] = `
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <p class="yfm-note-title">
    Outer title
  </p>
  <div class="yfm-note-content">
    <p>
      Outer content
    </p>
    <div class="yfm-note yfm-accent-info"
         note-type="info"
    >
      <p class="yfm-note-title">
        Inner title
      </p>
      <div class="yfm-note-content">
        <p>
          Inner content
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Alerts should render note with empty string title 1`] = `
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <div class="yfm-note-content">
    <p>
      Note content with empty string title
    </p>
  </div>
</div>
`;

exports[`Alerts should render note without title (notesAutotitle: false) 1`] = `
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <div class="yfm-note-content">
    <p>
      Note content without title
    </p>
  </div>
</div>
`;

exports[`Alerts should render note without title (notesAutotitle: true) 1`] = `
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <p class="yfm-note-title">
    Примечание
  </p>
  <div class="yfm-note-content">
    <p>
      Note content without title
    </p>
  </div>
</div>
`;

exports[`Alerts should render siblings notes 1`] = `
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <p class="yfm-note-title">
    Note title 1
  </p>
  <div class="yfm-note-content">
    <p>
      Note content 1
    </p>
  </div>
</div>
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <p class="yfm-note-title">
    Note title 2
  </p>
  <div class="yfm-note-content">
    <p>
      Note content 2
    </p>
  </div>
</div>
`;

exports[`Alerts should render simple note 1`] = `
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <p class="yfm-note-title">
    Note title
  </p>
  <div class="yfm-note-content">
    <p>
      Note content
    </p>
  </div>
</div>
`;

exports[`Alerts should render title with format 1`] = `
<div class="yfm-note yfm-accent-info"
     note-type="info"
>
  <p class="yfm-note-title">
    <em>
      Italic note title
    </em>
  </p>
  <div class="yfm-note-content">
    <p>
      Content
    </p>
  </div>
</div>
`;
